File: maximum-subarray.py
class Solution:
    # dynamic programming, memoization
    def maxSubArray(self, nums: List[int]) -> int:
        for i in range(1, len(nums)):
            nums[i] += nums[i - 1] if nums[i - 1] > 0 else 0

        return max(nums)

    # Kadane's algorithm: O(n)
    def maxSubArray(self, nums: List[int]) -> int:
        best_sum = -sys.maxsize
        current_sum = 0
        for num in nums:
            current_sum = max(num, current_sum + num)
            best_sum = max(best_sum, current_sum)
        
        return best_sum

    # Time Limit Exceeded
    def maxSubArray(self, nums: List[int]) -> int:
        def findSubArray(lrange, rrange, sums):
            if lrange == rrange:
                return sums

            left = findSubArray(lrange, rrange - 1, sums - nums[rrange])
            right = findSubArray(lrange + 1, rrange, sums - nums[lrange])

            return max(left, right, sums)
        
        return findSubArray(0, len(nums) - 1, sum(nums))

File: maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        for(int i=1; i<nums.length; i++)
            nums[i] += nums[i - 1] > 0? nums[i - 1]: 0;
        return maxArray(nums);
    }

    public int maxArray(int[] nums) {
        int max = nums[0];
        for(int num: nums) {
            max = Math.max(max, num);
        }
        return max;
    }

    public int maxSubArray(int[] nums) {
        int bestSum = Integer.MIN_VALUE;
        int currentSum = 0;
        for(int num: nums) {
            currentSum = Math.max(num, currentSum + num);
            bestSum = Math.max(bestSum, currentSum);
        }
        return bestSum;
    }
}